algebird-core/src/main/scala/com/twitter/algebird/immutable/BloomFilter.scala

```diff
@@ -0,0 +1,432 @@
+      var idx = 0
+      var found = true
+      while (idx < il.length && found) {
+        val i = il(idx)
+        if (!bits(i)) {
+          found = false
+        }
+        idx += 1
+      }
+      found
+    }
+
+    // use an approximation width of 0.05
+    override def size: Approximate[Long] =
+      BloomFilter.sizeEstimate(numBits, numHashes, width, 0.05)
+  }
+
+  implicit val monoid: Monoid[Hash] with BoundedSemilattice[Hash] =
+    new Monoid[Hash] with BoundedSemilattice[Hash] {
+      override val zero: Hash = Empty
+
+      /**
+       * Assume the bloom filters are compatible (same width and same hashing functions).  This
+       * is the union of the 2 bloom filters.
+       */
+      override def plus(left: Hash, right: Hash): Hash = left ++ right
+
+      override def sum(t: TraversableOnce[Hash]): Hash =
+        if (t.iterator.isEmpty) empty
+        else {
+          val iter = t.iterator
+          var bs = BitSet.newEmpty(0)
+          val hash = new Array[Int](numHashes)
+
+          while (iter.hasNext) {
+            iter.next() match {
+              case Empty => ()
+              case Item(item) =>
+                hashToArray(item, hash)
+                bs = bs.mutableAdd(hash)
+              case Instance(bitset) =>
+                val iter = bitset.iterator
+                while (iter.hasNext) {
+                  bs = bs.mutableAdd(iter.next())
+                }
+            }
+          }
+          if (bs.isEmpty) Empty else Instance(bs)
+        }
+
+      override def sumOption(t: TraversableOnce[Hash]): Option[Hash] =
+        if (t.iterator.isEmpty) None else Some(sum(t))
+    }
+
+  val aggregator: MonoidAggregator[A, Hash, Hash] = new MonoidAggregator[A, Hash, Hash] {
+    override val monoid: Monoid[Hash] = self.monoid
+
+    override def prepare(value: A): Hash = Item(value)
+
+    override def present(bf: Hash): Hash = bf
+  }
+
+  implicit val equiv: Equiv[Hash] = new Equiv[Hash] {
+    override def equiv(a: Hash, b: Hash): Boolean =
+      (a eq b) || a.toBitSet.equals(b.toBitSet)
+  }
+
+  /**
+   * Create a bloom filter with one item.
+   */
+  def create(item: A): Hash = Item(item)
+
+  /**
+   * Create a bloom filter with multiple items.
+   */
+  def create(data: A*): Hash = create(data.iterator)
+
+  /**
+   * Create a bloom filter with multiple items from an iterator
+   */
+  def create(data: Iterator[A]): Hash = monoid.sum(data.map(Item))
+
+  val empty: Hash = Empty
+
+  /**
+   * Attempts to create a new BloomFilter instance from a [[BitSet]]. Failure might occur
+   * if the BitSet has a maximum entry behond the BloomFilter expected size.
+   *
+   * This method will be helpfull on BloomFilter desirialization. Serialization is achieved
+   * through the serialization of the underlying [[BitSet]].
+   */
+  def fromBitSet(bitSet: BitSet): Try[Hash] =
+    if (bitSet.isEmpty) {
+      Success(empty)
+    } else {
+      if (bitSet.toSet.max > width)
```

Comment:
This is going to be pretty slow. Can we use reverseIterator.next() from bitSet instead?

Reply:
You are right! we totally can! 🤦‍♂️ give me 1 sec

Reply:
I overlooked that... that was embarrassing lol.